spec/models/printables/group_spec.rb

```diff
@@ -82,75 +61,142 @@
+      allow(Rails.configuration).to receive(:pmb_uri).and_return(uri)
+    end
+
+    context 'when using v1' do
+      let(:uri) { 'http://localhost:10000/v1' }
+      let(:post_headers) { { 'Accept' => 'application/vnd.api+json', 'Content-Type' => 'application/vnd.api+json' } }
+      it 'sends the right message to PMB' do
+        request =
+          stub_request(:post, "#{uri}/print_jobs")
+            .with(
+              body: {
+                data: {
+                  type: 'print_jobs',
+                  attributes: {
+                    printer_name: 'printer1',
+                    label_template_id: 1,
+                    labels: {
+                      body: [
+                        { label: { barcode: '2', barcode2d: '2', top_line: '', middle_line: nil, bottom_line: '' } },
+                        { label: { barcode: '1', barcode2d: '1', top_line: '', middle_line: nil, bottom_line: '' } }
+                      ]
+                    }
+                  }
+                }
+              }.to_json,
+              headers: post_headers
+            )
+            .to_return(status: 200, body: <<~RESPONSE, headers: { 'Content-Type' => 'application/json' })
+                {"data":{"type":"print_jobs","attributes":{"printer_name":"printer1","label_template_id":1,"labels":{"body":[{"label":{"barcode":"2","barcode2d":"2","top_line":"","middle_line":null,"bottom_line":""}},{"label":{"barcode":"1","barcode2d":"1","top_line":"","middle_line":null,"bottom_line":""}}]}}}}
+                RESPONSE
+
+        group.print(config, 'user1')
+        expect(request).to have_been_made
+      end
+
+      context 'when an asset does not have barcode' do
+        let(:asset1) { create(:asset, facts: facts1, barcode: nil) }
+        it 'does not print it' do
+          request =
+            stub_request(:post, "#{uri}/print_jobs")
+              .with(
+                body: {
+                  data: {
+                    type: 'print_jobs',
+                    attributes: {
+                      printer_name: 'printer1',
+                      label_template_id: 1,
+                      labels: {
+                        body: [
+                          { label: { barcode: '2', barcode2d: '2', top_line: '', middle_line: nil, bottom_line: '' } }
+                        ]
+                      }
+                    }
+                  }
+                }.to_json,
+                headers: post_headers
+              )
+              .to_return(status: 200, body: <<~RESPONSE, headers: { 'Content-Type' => 'application/vnd.api+json' })
+                  {"data":{"type":"print_jobs","attributes":{"printer_name":"printer1","label_template_id":1,"labels":{"body":[{"label":{"barcode":"1","barcode2d":"1","top_line":"","middle_line":null,"bottom_line":""}}]}}}}
+                  RESPONSE
+
+          group.print(config, 'user1')
+          expect(request).to have_been_made
```

Comment:
Don't quite follow this test, `barcode: 2` is stubbed, but the response barcode is different, even though the test name says when no barcode is provided. Should it still return a 200?

Reply:
Its in group spec, so in this case some of the assets still have barcodes, and they get printed off. In this case its only asset1 that has no barcode.

Reply:
Cool - makes sense, thanks

Reply:
And we dont mind it returns a 200 even though some failed?

Reply:
We're not really in control of the return response here, that's from PMB. So as far as PMB it concerned, everything *is* fine, as we've only sent it the barcodes that work.

In terms of what we do, I've followed the existing behaviour, where missing barcodes are ignored silently. This behaviour was explicitly tested, so is intentional, although the reasons weren't documented. I suspect this is because printing is coupled to the other actions, such that barcodes are automatically printed as required. These steps handle a variety of assets, some of which might not even have/need barcodes.

We've got a story in the backlog to try and reduce some of this coupling a bit, where we could probably be a bit more explicit about what our expectations are.